scalability_and_performance/understanding-node-observability-operator.adoc

```diff
@@ -0,0 +1,363 @@
+====
+You can request only one profiling query at any point of time.
+====
+
+.Prerequisites
+* The Node Observability Operator has been installed.
+* The `NodeObservability` custom resource (CR) has been created.
+* You have access to the cluster with `cluster-admin` privileges.
+
+.Procedure
+
+. Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:
++
+[source,yaml]
+----
+apiVersion: nodeobservability.olm.openshift.io/v1alpha1
+kind: NodeObservabilityRun
+metadata:
+  name: nodeobservabilityrun
+spec:
+  nodeObservabilityRef:
+    name: cluster
+----
+
+. Run the `NodeObservabilityRun` to trigger the profiling:
++
+[source,terminal]
+----
+$ oc apply -f nodeobservabilityrun.yaml
+----
+
+. Review the status of the `NodeObservabilityRun`:
++
+[source,terminal]
+----
+$ oc get nodeobservabilityrun -o yaml --watch
+----
+
++
+.Example output
+[source,terminal]
+----
+apiVersion: nodeobservability.olm.openshift.io/v1alpha1
+kind: NodeObservabilityRun
+metadata:
+  annotations:
+    kubectl.kubernetes.io/last-applied-configuration: |
+      {"apiVersion":"nodeobservability.olm.openshift.io/v1alpha1","kind":"NodeObservabilityRun","metadata":{"annotations":{},"name":"nodeobservabilityrun-sample","namespace":"node-observability-operator"},"spec":{"nodeObservabilityRef":{"name":"cluster"}}}
+  creationTimestamp: "2022-07-07T14:57:34Z"
+  generation: 1
+  name: nodeobservabilityrun-sample
+  namespace: node-observability-operator
+  ownerReferences:
+  - apiVersion: nodeobservability.olm.openshift.io/v1alpha1
+    kind: NodeObservability
+    name: cluster
+    uid: d544a832-2100-4574-8eee-aec70544c5ed
+  resourceVersion: "297555"
+  uid: 7b77c188-2e7f-4796-9c3d-170a0f873152
+spec:
+  nodeObservabilityRef:
+    name: cluster
+status:
+  agents:
+  - ip: 10.217.0.68
+    name: node-observability-ds-wvrxc
+    port: 8443
+  conditions:
+    conditions:
+    - lastTransitionTime: "2022-07-07T14:57:34Z"
+      message: Ready to start profiling
+      reason: Ready
+      status: "True"
+      type: Ready
+    - lastTransitionTime: "2022-07-07T14:58:10Z"
+      message: Profiling query done
+      reason: Finished
+      status: "True"
+      type: Finished
+  finishedTimestamp: "2022-07-07T14:58:10Z"
+  startTimestamp: "2022-07-07T14:57:35Z"
+----
+
++
+Profiling query is complete when the status `True` and type `Finished`.
+
+. Run the following bash script to store the profiling data to `/run/node-observability` path:
++
+[source,bash]
+----
+for i in `oc get nodeobservabilityrun.nodeobservability.olm.openshift.io/nodeobservabilityrun -o yaml | yq .status.agents[].name | cut -d\" -f2`
+  do
+  echo $i
+  list=`oc exec $i -c node-observability-agent -- bash -c "ls /run/node-observability/*.pprof"`
+  for j in $list
+    do
+    k=`echo $j|cut -d\/ -f3`
+    mkdir -p /tmp/$i
+    echo copying $k to /tmp/$i/$k
+    kubectl exec $i -c node-observability-agent -- cat $j > /tmp/$i/$k
+  done
+done
```

Comment:
Smaller and stricter bash version of this script:
```
for a in $(oc get nodeobservabilityrun nodeobservabilityrun -o yaml | yq .status.agents[].name); do
  echo "agent ${a}"
  mkdir -p "/tmp/${a}"
  for p in $(oc exec "${a}" -c node-observability-agent -- bash -c "ls /run/node-observability/*.pprof"); do
    f="$(basename ${p})"
    echo "copying ${f} to /tmp/${a}/${f}"
    oc exec "${a}" -c node-observability-agent -- cat "${p}" > "/tmp/${a}/${f}"
  done
done
```

Reply:
Manually tested:
```
$ ./script.sh
agent node-observability-ds-zp24r
copying crio-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof to /tmp/node-observability-ds-zp24r/crio-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof
copying kubelet-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof to /tmp/node-observability-ds-zp24r/kubelet-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof
agent node-observability-ds-szdf4
copying crio-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof to /tmp/node-observability-ds-szdf4/crio-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof
copying kubelet-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof to /tmp/node-observability-ds-szdf4/kubelet-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof
agent node-observability-ds-wct6p
copying crio-865dc234-2b0d-444e-bd18-44333355ed77.pprof to /tmp/node-observability-ds-wct6p/crio-865dc234-2b0d-444e-bd18-44333355ed77.pprof
copying kubelet-865dc234-2b0d-444e-bd18-44333355ed77.pprof to /tmp/node-observability-ds-wct6p/kubelet-865dc234-2b0d-444e-bd18-44333355ed77.pprof

$ ls -ltr /tmp/node-observability-ds-*
/tmp/node-observability-ds-zp24r:
total 32
-rw-rw-r--. 1 alebedev alebedev  7060 Jul 11 12:11 crio-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof
-rw-rw-r--. 1 alebedev alebedev 22511 Jul 11 12:11 kubelet-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof

/tmp/node-observability-ds-szdf4:
total 32
-rw-rw-r--. 1 alebedev alebedev  6635 Jul 11 12:11 crio-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof
-rw-rw-r--. 1 alebedev alebedev 23364 Jul 11 12:11 kubelet-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof

/tmp/node-observability-ds-wct6p:
total 28
-rw-rw-r--. 1 alebedev alebedev  3345 Jul 11 12:12 crio-865dc234-2b0d-444e-bd18-44333355ed77.pprof
-rw-rw-r--. 1 alebedev alebedev 21965 Jul 11 12:12 kubelet-865dc234-2b0d-444e-bd18-44333355ed77.pprof
```

Reply:
Fixed.